### PR TITLE
Controller map supplier

### PR DESCRIPTION
**Pull request by thumbnail**
This allows for dynamic rebinding of the controllers to the players.  

An example for subra that'll increase the id of the assigned robot by 1 every 500ms would be:
```java
Publisher<Controller.MapSupplier> mappingPublisher = Flux.interval(Duration.ofMillis(500))
  .map(tick -> () -> ImmutableMap.of(
    new PlayerIdentity(tick.intValue(), TeamColor.BLUE),
    ImmutableSet.of(new ControllerIdentity(1)))

new Zosma(
  new ControllerPublisher<>(2, Duration.ofMillis(100), new XboxJamepadAdapter(), mappingPublisher),
  new ControllerDeducer<>(new ControllerHandler<>(), AveragePlayerCommand::new),
  System.out::println
).run();
```
Note that the the flux returns a lambda, so it acts as an `Controller.MapSupplier`.

### CODE REVIEW

File: src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java [hunk 1]
```diff
@@ -1,87 +1,86 @@
...
+  public void subscribe(final Subscriber<? super Frame<C, I>> s) {
+    Flux.combineLatest(setSupplierPublisher, mapSupplierPublisher, Frame::new).subscribe(s);
   }
 
-  /**
-   * The Class Frame.
-   *
-   * @author Jeroen de Jong
-   */
   @Value
```
Comment (thumbnail) [inline src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java]: This should be `@AllArgsConstructor`.

File: src/main/java/io/leonis/zosma/ipc/peripheral/ControllerSetPublisher.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            // Loop all controllers
+            IntStream.range(0, manager.getNumControllers())
+                // grab the states
+                .mapToObj(manager::getControllerIndex)
+                // ignore any disconnected controller
+                .filter(ControllerIndex::isConnected)
+                // map to custom state and collect as a set.
+                .map(adapter)
+                .collect(Collectors.toSet()))
+        .map(Frame::new)
```
Comment (thumbnail) [inline src/main/java/io/leonis/zosma/ipc/peripheral/ControllerSetPublisher.java]: Frame can be replaced with a lambda

File: src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java [hunk 3]
```diff
@@ -1,87 +1,86 @@
...
-  @Value
-  private class Frame implements Controller.SetSupplier<C> {
-    /**
-     * The {@link Set} of {@link Controller controllers} captured in this frame.
-     */
-    private Set<C> controllerSet;
+  @AllArgsConstructor
+  static final class Frame<C extends Controller, I extends Identity>
+      implements Controller.SetSupplier<C>, Controller.MapSupplier<I> {
+    @lombok.experimental.Delegate
```
Comment (romni) [inline src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java:81]: these should be `@Delegate`

Comment (thumbnail) [reply src/main/java/io/leonis/zosma/ipc/peripheral/ControllerPublisher.java:81]: Can't be done due to the import of `lombok.*` or something. Please check for yourself locally whether you can get it to work though :P

**Pull request closed without merging**